src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/cas3/DomainEventServiceTest.kt

```diff
@@ -1731,4 +1516,220 @@ class DomainEventServiceTest {
+            it.data == objectMapper.writeValueAsString(domainEventToSave.data)
+        },
+      )
+    }
+    verify(exactly = 0) {
+      mockHmppsTopic.snsClient.publish(any())
+    }
+  }
+
+  @Test
+  fun `Should throw error when savePersonDepartureUpdatedEvent fail to publish and save event in DB`() {
+    val id = UUID.fromString("c3b98c67-065a-408d-abea-a252f1d70981")
+    val applicationId = UUID.fromString("a831ead2-31ae-4907-8e1c-cad74cb9667b")
+    val occurredAt = OffsetDateTime.parse("2023-02-01T14:03:00+00:00")
+    val crn = "CRN"
+    val domainEventToSave = createCAS3DepartureUpdatedDomainEvent(id, applicationId, crn, occurredAt)
+    val bookingEntity = createTemporaryAccommodationPremisesBookingEntity()
+    val mockHmppsTopic = mockk<HmppsTopic>()
+    every { domainEventRepositoryMock.save(any()) } answers { it.invocation.args[0] as DomainEventEntity }
+    every { hmppsQueueServiceMock.findByTopicId("domainevents") } returns mockHmppsTopic
+    every { domainEventBuilderMock.buildDepartureUpdatedDomainEvent(any()) } returns domainEventToSave
+    every { mockHmppsTopic.arn } returns "arn:aws:sns:eu-west-2:000000000000:domain-events"
+    every { mockHmppsTopic.snsClient.publish(any()) } throws InternalErrorException("Unexpected exception")
+
+    assertThatExceptionOfType(InternalErrorException::class.java)
+      .isThrownBy { domainEventService.savePersonDepartureUpdatedEvent(bookingEntity) }
+
+    verify(exactly = 1) {
+      domainEventRepositoryMock.save(
+        match {
+          it.id == domainEventToSave.id &&
+            it.type == DomainEventType.CAS3_PERSON_DEPARTURE_UPDATED &&
+            it.crn == domainEventToSave.crn &&
+            it.occurredAt.toInstant() == domainEventToSave.occurredAt &&
+            it.data == objectMapper.writeValueAsString(domainEventToSave.data)
+        },
+      )
+    }
+    verify(exactly = 1) {
+      mockHmppsTopic.snsClient.publish(
+        match {
+          val deserializedMessage = objectMapper.readValue(it.message, SnsEvent::class.java)
+
+          deserializedMessage.eventType == "accommodation.cas3.person.departed.updated" &&
+            deserializedMessage.version == 1 &&
+            deserializedMessage.description == "Person has updated departure date of Transitional Accommodation premises" &&
+            deserializedMessage.detailUrl == "http://api/events/cas3/person-departure-updated/$id" &&
+            deserializedMessage.occurredAt.toInstant() == domainEventToSave.occurredAt &&
+            deserializedMessage.additionalInformation.applicationId == applicationId &&
+            deserializedMessage.personReference.identifiers.any { it.type == "CRN" && it.value == domainEventToSave.data.eventDetails.personReference.crn } &&
+            deserializedMessage.personReference.identifiers.any { it.type == "NOMS" && it.value == domainEventToSave.data.eventDetails.personReference.noms }
+        },
+      )
+    }
+  }
+
```

Comment:
It would also be good to have a test to make sure that this event is saved but not emitted to SNS if it's disabled in the configuration (see `saveReferralSubmittedEvent persists event, but does not emit event to SNS when event is disabled` as an example).

Reply:
oh yes, its your recent change. I will add test for it, thanks for reminding me

Reply:
done